### PR TITLE
Include image_enrichment module and supporting code

### PR DESCRIPTION
**Pull request by jrmontag**
The new image_enrichment module includes an enrichment class definition
for applying labels to attached Tweet images based on [an open source
implementation](https://keras.io/applications/) of the VGG16 network
architecture. Some of the methods of this class could be abstracted
into a generic base class if/when the need arises, but for now are
not.

The setup is a bit more involved for this enrichment because of the
list of library dependencies, and the one-time downloads of the network
weights and class labels. The included bash script (and requirements
file) is an attempt to make this relatively straightforward for the
user.

-------

Hi @tw-ddis/ddis team, I'd appreciate some eyes on this PR. As the commit above explains, its adding the image labeling prototype from earlier this year into the pipeline as an enrichment (via the tools repo). 

If you'd like to test the branch locally, here's how I tested on another machine:
1. create a new python 3 env and activate it
2. clone Gnip-Analysis-Tools and then `git checkout img-label-enrichment`
3. install this branch as an env library from the root of the repo: `pip install -e .`
4. go to `gnip_analysis_config/enrichments/` and run `bash image-build.sh` 

Then, in a separate project directory you can test the whole thing with some Tweet data and the `pip install`'d Analysis Pipeline code. As in the GAP readme, create a small module, `test_img_enrichment.py`:

```python
from gnip_analysis_config.enrichments import image_enrichment

enrichment_class_list = image_enrichment.image_enrichments_list
```

And then send some newline-delimited Tweet data through it to inspect the results:
```bash
(env) [project dir] $ cat tweet_data.json | tweet_enricher.py -c test_img_enrichment.py | jq '.enrichments' -c
```

**Future work**
Things that I chose not to bite off for this first PR, but will open as issues for reasonable next-steps:
- replace explicit Activity Streams dict access with the new Tweet parser library
- support conda envs in the bash setup script
- account for >1 image attached to Tweet

### CODE REVIEW

File: gnip_analysis_config/enrichments/image_enrichment.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+from keras.applications.vgg16 import decode_predictions as vgg16_decode_predictions
+
+
+class MyEnrichment(enrichment_base.BaseEnrichment):
+    def enrichment_value(self, tweet):
+        return "my_test_enrichment_value"
+
+
+class ImageLabelVGG16(enrichment_base.BaseEnrichment):
+    def __init__(self):
```
Comment (binaryaaron) [inline gnip_analysis_config/enrichments/image_enrichment.py]: Probably need a docstring here, possibly links to Keras and the demo posts from there?

File: gnip_analysis_config/enrichments/image_enrichment.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+from keras.applications.vgg16 import preprocess_input as vgg16_preprocess_input
+from keras.applications.vgg16 import decode_predictions as vgg16_decode_predictions
+
+
+class MyEnrichment(enrichment_base.BaseEnrichment):
+    def enrichment_value(self, tweet):
+        return "my_test_enrichment_value"
+
+
+class ImageLabelVGG16(enrichment_base.BaseEnrichment):
```
Comment (binaryaaron) [inline gnip_analysis_config/enrichments/image_enrichment.py]: Eventually might want to let the user supply a model or provide an option for a model that isn't VGG.

File: gnip_analysis_config/enrichments/image_enrichment.py (hunk 1 above)
Comment (jrmontag) [inline gnip_analysis_config/enrichments/image_enrichment.py]: Yep, good call. Will add.

File: gnip_analysis_config/enrichments/image_enrichment.py (hunk 2 above)
Comment (jrmontag) [inline gnip_analysis_config/enrichments/image_enrichment.py]: Yes, there may (or may not) be a desire to use other pre-trained models in the future, so I was trying not to prematurely modularize :) But, in the event that there is such a desire, the first step would likely be to factor out something like a generic `ImageLabel` base class that has all the useful methods, and then enable something like a model-specific class (maybe reclaim the `ImageLabelVGG16` name) that can inherit from the base class but have a different `self.model`. I think the [NTLK tokenizers](https://github.com/tw-ddis/Gnip-Analysis-Tools/blob/master/gnip_analysis_config/enrichments/nltk_enrichment.py#L14) have a good example of this.

My initial thought was to make that split if/when the time comes. But if it seems important, I can do it in this PR.

File: gnip_analysis_config/enrichments/image_enrichment.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+import numpy as np
+from PIL import Image
+import requests
+from keras.preprocessing import image as keras_image
+from keras.applications.vgg16 import VGG16
+from keras.applications.vgg16 import preprocess_input as vgg16_preprocess_input
+from keras.applications.vgg16 import decode_predictions as vgg16_decode_predictions
+
+
+class MyEnrichment(enrichment_base.BaseEnrichment):
```
Comment (jrmontag) [inline gnip_analysis_config/enrichments/image_enrichment.py]: Leaving this class definition in here was a mistake. Will remove!

File: gnip_analysis_config/enrichments/image_enrichment.py (hunk 2 above)
Comment (binaryaaron) [inline gnip_analysis_config/enrichments/image_enrichment.py]: Not important now, just a thought for the future. If there is any broad interest, having interchangeable image enrichments might be nice.

**Comment by jrmontag** [comment]
I think I addressed all of the above issues:
- removed the extraneous demo class
- factorized out a base class (I tried to resist, but the VGG class is so nice and tidy now...)
- added docstrings to the base class pointing to the Keras docs from which all of this was inspired

Looking forward to other comments.

File: gnip_analysis_config/enrichments/image-build.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+if [[ -z ${VIRTUAL_ENV} ]]; then
+    echo
+    echo "Please build a Python3 virtualenv prior to running this script."
+    echo 
+    exit 0
+fi
+
+echo "$(date +%Y-%m-%d\ %H:%M:%S) -- installing python libraries (takes a bit the first time)"
+${VIRTUAL_ENV}/bin/pip install -U pip > /dev/null
+${VIRTUAL_ENV}/bin/pip install -r image-reqs.txt > /dev/null
```
Comment (jeffakolb) [inline gnip_analysis_config/enrichments/image-build.sh]: I think you want a '-U' here. I had an older version of keras installed, which did not have the 'applications' subpackage.

Comment (jeffakolb) [inline gnip_analysis_config/enrichments/image-build.sh]: @jrmontag Also, per our conversation, I needed to delete ~/.keras/keras.json to remove the selection of the theano backend.

Comment (jrmontag) [reply gnip_analysis_config/enrichments/image-build.sh]: Thanks @jeffakolb. I took a middle ground and specified minimum (recent) versions of tensorflow and keras. In particular, this should address the breaking changes introduce with Keras 2.0. 

I'm going to wait on specific usage guidance regarding the keras.json file for two reasons:
1) in Keras 2.0 (now required for use of this module), the default backend is tensorflow and I'd like to see how a Theno-based keras.json handles that case. 
2) In principle, the backend should be an abstraction - managed by Keras - that doesn't affect the user. 

So I'm happy to address that again after we have more explicit examples of current and expected behavior.
